### ISSUE TITLE
Certain sounds are registered incorrectly on client in multiplayer

### ISSUE THREAD

**Issue by noodlecollie** [FWGS/xash3d-fwgs#35]
I understand that the multiplayer mode is still largely untested and a bit buggy, so I was hoping to find out some more information about the following:

Some of my weapon sounds, when tested in multiplayer, ended up being completely different sound files to what was expected. I've narrowed it down to the fact that `S_FreeSound` is being called prematurely on the list of sounds created on precache, so sounds that are loaded in later take the place of the precached sounds in the `s_knownSfx` list.

The reason that sounds are being freed is that `S_BeginRegistration`/`S_EndRegistration` are being called more than once. Each time begin is called, `s_registration_sequence` is incremented. When end is called, any sounds whose sequence number doesn't match the current `s_registration_sequence` are freed. This seems to be what happens to all of my precached sounds.

The first time begin is called is on receipt of an `svc_resourcelist` message:

```
libxash.dylib!S_BeginRegistration (/Users/vesper/Documents/afterburner-game/dependencies/afterburner-engine/engine/client/s_load.c:293)
libxash.dylib!CL_PrecacheResources (/Users/vesper/Documents/afterburner-game/dependencies/afterburner-engine/engine/client/cl_main.c:3161)
libxash.dylib!CL_BatchResourceRequest (/Users/vesper/Documents/afterburner-game/dependencies/afterburner-engine/engine/client/cl_parse.c:543)
libxash.dylib!CL_StartResourceDownloading (/Users/vesper/Documents/afterburner-game/dependencies/afterburner-engine/engine/client/cl_parse.c:632)
libxash.dylib!CL_ParseResourceList (/Users/vesper/Documents/afterburner-game/dependencies/afterburner-engine/engine/client/cl_parse.c:1682)
libxash.dylib!CL_ParseServerMessage (/Users/vesper/Documents/afterburner-game/dependencies/afterburner-engine/engine/client/cl_parse.c:2274)
libxash.dylib!CL_ReadNetMessage (/Users/vesper/Documents/afterburner-game/dependencies/afterburner-engine/engine/client/cl_main.c:2766)
libxash.dylib!CL_ReadPackets (/Users/vesper/Documents/afterburner-game/dependencies/afterburner-engine/engine/client/cl_main.c:2802)
libxash.dylib!Host_ClientFrame (/Users/vesper/Documents/afterburner-game/dependencies/afterburner-engine/engine/client/cl_main.c:3524)
libxash.dylib!Host_Frame (/Users/vesper/Documents/afterburner-game/dependencies/afterburner-engine/engine/common/host.c:544)
libxash.dylib!Host_RunFrame (/Users/vesper/Documents/afterburner-game/dependencies/afterburner-engine/engine/common/host_state.c:144)
libxash.dylib!COM_Frame (/Users/vesper/Documents/afterburner-game/dependencies/afterburner-engine/engine/common/host_state.c:194)
libxash.dylib!Host_Main (/Users/vesper/Documents/afterburner-game/dependencies/afterburner-engine/engine/common/host.c:1034)
bond!Sys_Start() (/Users/vesper/Documents/afterburner-game/dependencies/afterburner-engine/game_launch/game.cpp:134)
bond!main (/Users/vesper/Documents/afterburner-game/dependencies/afterburner-engine/game_launch/game.cpp:146)
```

And the second time appears to be because of the condition `cls.signon == SIGNONS && cl.lastresourcecheck < host.realtime` in `CL_ReadPackets`:

```
libxash.dylib!S_BeginRegistration (/Users/vesper/Documents/afterburner-game/dependencies/afterburner-engine/engine/client/s_load.c:293)
libxash.dylib!CL_PrecacheResources (/Users/vesper/Documents/afterburner-game/dependencies/afterburner-engine/engine/client/cl_main.c:3161)
libxash.dylib!CL_BatchResourceRequest (/Users/vesper/Documents/afterburner-game/dependencies/afterburner-engine/engine/client/cl_parse.c:543)
libxash.dylib!CL_ReadPackets (/Users/vesper/Documents/afterburner-game/dependencies/afterburner-engine/engine/client/cl_main.c:2819)
libxash.dylib!Host_ClientFrame (/Users/vesper/Documents/afterburner-game/dependencies/afterburner-engine/engine/client/cl_main.c:3524)
libxash.dylib!Host_Frame (/Users/vesper/Documents/afterburner-game/dependencies/afterburner-engine/engine/common/host.c:544)
libxash.dylib!Host_RunFrame (/Users/vesper/Documents/afterburner-game/dependencies/afterburner-engine/engine/common/host_state.c:144)
libxash.dylib!COM_Frame (/Users/vesper/Documents/afterburner-game/dependencies/afterburner-engine/engine/common/host_state.c:194)
libxash.dylib!Host_Main (/Users/vesper/Documents/afterburner-game/dependencies/afterburner-engine/engine/common/host.c:1034)
bond!Sys_Start() (/Users/vesper/Documents/afterburner-game/dependencies/afterburner-engine/game_launch/game.cpp:134)
bond!main (/Users/vesper/Documents/afterburner-game/dependencies/afterburner-engine/game_launch/game.cpp:146)
```

Is there a reason that sound registration should be performed more than once, or is this incorrect behaviour?

**Comment by a1batross** [comment]
It's probably incorrect behavior.

Looks like engine tries to register sound after reslist, because "thinks" that your sound file as resource, is missing(RES_WASMISSING flag is set). Check CL_EstimateNeededResources call, in CL_ReadPackets.

**Comment by x6herbius** [comment]
OK, I've narrowed it down: it appears I was forgetting to precache a sound (gun pickup), so when the player spawned and picked up their first gun, this sound was sent in a sound message to the client so that it would be loaded. This caused `cl.resourcesneeded` to be non-empty, so sounds were registered again to pick up this non-precached sound. However, running sound registration again meant that all sounds from the previous registration were wiped from the list.

Obviously a simple fix would be to make sure the offending sound is precached. However, I'm still confused as to why forgetting to precache a sound would cause the rest of the correctly precached sounds to be discarded from the known SFX list. What would the best fix for that be?

**Comment by a1batross** [comment]
@x6herbius 

Precaching is very important thing in Quake. It helps engine build list of important files, which should be sent to client and preloaded(to avoid stutterring). So don't forget to precache your files. 

However, Xash3D supports hot precaching. It's possible that hot precaching is broken with customization, which allows any client send it's files to server and server will share it to other players. It's hardly allowed only for custom decals though.

**Comment by x6herbius** [comment]
I guess my question is, should hot precaching be triggering a content registration if this wipes the list of previously registered content? Without knowing all that much about the intent behind the system, it seems to me like items that need hot precaching should be added to the end of the existing list, rather than causing a re-registration.

**Comment by a1batross** [comment]
Was fixed way back in 2022 in 7013d447ca4bcd94b095489cb338277e870444ba
